docs/adrs/002-Retry_endpoint.md

```diff
@@ -0,0 +1,81 @@
+# Retry Endpoint
+
+Date: 2021-04-27
+
+The current requirement for the Batch/Bulk UI is to be able to retry jobs. Using a job id, it should be possible to get information from the database.
+
+The current implementation of retry is to run jobs using the `run_job` or `run_job_batch` endpoint. This is not adequate due to the following deficiencies:
+
+
+* Lack of book-keeping for the parent job and the child job relationship: Launching a new job will not update the relationship between those jobs
+* Lack of book-keeping for number of retries for a particular job / set of job inputs
+* Lack of ability to launch multiple jobs using the `run_job_batch` endpoint without creating a new parent job
+* Lack of ability to ensure that the proper catalog version /git commit of an app is used from the front end based on a tag, such as "beta/dev/release"
+* Lack of ability to specify which retries succeeded and which ones failed during submit time. 
+
+* Code is split more than is necessary
+
+## Author(s)
+
+@bio-boris
+
+## Status
+
+Pending
+
+## Alternatives Considered
+
+* Not book-keeping, or doing minimal book-keeping and calling run_job multiple times
+* Re-writing run_job/run_job_batch to address the aforementioned deficiencies
+* Creating a retry endpoint dedicated to addressing book-keeping and job launching features
+
+## Decision Outcome
+
+Pending
+
+* Link to spec file with inputs and outputs for the retry endpoint goes here
+* Link to Jira Ticket with business logic documentation for success and cancel cases
+
+### Possible additional things to think about
+* Creating sets of objects and what to do at the end of a batch run
+* What to do about a set if a child task fails during processing 
+* Convenience endpoints that operate on the parent_job_id or list of child job ids may be out of scope (e.g. cancel all jobs with a certain status)
+
+## Consequences
+
+Pending decision
+
+## Pros and Cons of the Alternatives
+
+### Not book-keeping, or doing minimal book-keeping and calling `run_job` multiple times
+* `+` Can re-use existing endpoints without any additional work
+* `+` Less api endpoints to manage
+* `-` Issues on re-rendering/regenerating a cell based on just the job record
+* `-` Loss of information about job runs, and ability to infer relationships between parents and child jobs. 
+* `-` Loss of control of jobs, such as the ability to restrict a job's running
+* `-` Wrong version of app will run if the app was updated after job completion, and a version tag rather than a git commit was provided
+* `-` Increase complexity of `run_job*` methods
+ 
+### Re-writing `run_job` to address the aforementioned deficiencies without refactoring
+* `+` Solves most requirements, but 
+* `-` Adds more complexity to `run_job` methods
+* `-` Increase difficulty in maintaining and testing `run_job` method
+* `-` Wrong version of app will run if the app was updated after job completion, and a version tag rather than a git commit was provided
```

Comment:
what happens with this at the moment -- does the job record store the version tag? the git commit? both?

Reply:
The job record stores the git commit.

Reply:
So this point about the version tag is not really relevant to the discussion.

Reply:
Yes, but the job is currently submitted via the code cell/narrative and that doesn't always store git commit. It stores "Dev|Beta|release|git commit". Submitting the job again with just the info from the browser has a chance to choose the wrong commit.

Reply:
Another possible issue here is that there's multiple ways to interpret "version" of an app.
If the app's in a module with the release tag, it has at least one semantic version. There should also be a git commit of what's deployed.
If it's either in beta or dev, there's no official release, so no semantic version. There's always a git commit, though.

It might be best to always submit a git commit for the module, maybe?

Reply:
> It might be best to always submit a git commit for the module, maybe?
+1

Reply:
should have known better than to think that the narrative would always act in the same way! 😖